From 0222d289f9584e170aec2a49080fffefa2f39dff Mon Sep 17 00:00:00 2001
From: Bernd Steinhauser <berniyh@exherbo.org>
Date: Thu, 10 Jul 2014 17:39:47 +0200
Subject: [PATCH 1/2] Update exif tag information for Panasonic cameras.

The tag information for existing tags has been updated
and new tags have been added, based on the information available at
http://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/Panasonic.html
and the output of a Panasonic G6.
---
 src/panasonicmn.cpp | 382 ++++++++++++++++++++++++++++++++++++++++++++++------
 1 file changed, 339 insertions(+), 43 deletions(-)

diff --git a/src/panasonicmn.cpp b/src/panasonicmn.cpp
index cbfe7bd..b962ad8 100644
--- a/src/panasonicmn.cpp
+++ b/src/panasonicmn.cpp
@@ -25,6 +25,17 @@
              Gilles Caulier (gc) <caulier dot gilles at gmail dot com>
   History:   11-Jun-04, ahu: created
   Credits:   See header file
+
+  The following tags have not yet been implemented, not sure how useful they would be:
+    0x003b  TextStamp   int16u  1 = Off 2 = On
+    0x003e  TextStamp   int16u  1 = Off 2 = On
+    0x004d AFPointPosition     rational64u[2]  (X Y coordinates of primary AF area center, in the range 0.0 to 1.0)
+    0x0059 Transform undef[4]   (decoded as two 16-bit signed integers)
+        '-1 1' = Slim Low '-3 2' = Slim High '0 0' = Off '1 1' = Stretch Low '3 2' = Stretch High
+    0x004e FaceDetInfo See http://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/Panasonic.html#FaceDetInfo
+    0x0061 FaceRecInfo See http://www.sno.phy.queensu.ca/~phil/exiftool/TagNames/Panasonic.html#FaceRecInfo
+    0x8008  TextStamp   int16u  1 = Off 2 = On
+    0x8009  TextStamp   int16u  1 = Off 2 = On
  */
 // *****************************************************************************
 #include "rcsid_int.hpp"
@@ -69,7 +80,9 @@ namespace Exiv2 {
         {  8, N_("Flash")           },
         { 10, N_("Black and white") },
         { 11, N_("Manual")          },
-        { 11, N_("Manual")          }           // To silence compiler warning
+        { 12, N_("Shade")           },
+        { 13, N_("Kelvin")          },
+        { 13, N_("Kelvin")          }           // To silence compiler warning
     };
 
     //! FocusMode, tag 0x0007
@@ -78,21 +91,27 @@ namespace Exiv2 {
         {  2, N_("Manual")             },
         {  4, N_("Auto, focus button") },
         {  5, N_("Auto, continuous")   },
-        {  5, N_("Auto, continuous")   }        // To silence compiler warning
+        {  6, N_("AF-S")               },
+        {  7, N_("AF-C")               },
+        {  8, N_("AF-F")               },
+        {  8, N_("AF-F")               }        // To silence compiler warning
     };
 
     //! ImageStabilizer, tag 0x001a
     extern const TagDetails panasonicImageStabilizer[] = {
         {  2, N_("On, Mode 1") },
         {  3, N_("Off")        },
-        {  4, N_("On, Mode 2") }
+        {  4, N_("On, Mode 2") },
+        {  5, N_("Panning") },
+        {  6, N_("On, Mode 3") }
     };
 
     //! Macro, tag 0x001c
     extern const TagDetails panasonicMacro[] = {
         {   1, N_("On")         },
         {   2, N_("Off")        },
-        { 257, N_("Tele-macro") }
+        { 257, N_("Tele-macro") },
+        { 513, N_("Macro Zoom") }
     };
 
     //! ShootingMode, tag 0x001f and SceneMode, tag 0x8001
@@ -113,6 +132,8 @@ namespace Exiv2 {
         { 13, N_("Panning")                        },
         { 14, N_("Simple")                         },
         { 15, N_("Color effects")                  },
+        { 16, N_("Self Portrait")                  },
+        { 17, N_("Economy")                        },
         { 18, N_("Fireworks")                      },
         { 19, N_("Party")                          },
         { 20, N_("Snow")                           },
@@ -132,12 +153,53 @@ namespace Exiv2 {
         { 34, N_("Intelligent ISO")                },
         { 36, N_("High speed continuous shooting") },
         { 37, N_("Intelligent auto")               },
+        { 39, N_("Multi-aspect")                   },
+        { 41, N_("Transform")                      },
+        { 42, N_("Flash Burst")                    },
+        { 43, N_("Pin Hole")                       },
+        { 44, N_("Film Grain")                     },
+        { 45, N_("My Color")                       },
+        { 46, N_("Photo Frame")                    },
+        { 48, N_("Movie")                          },
+        { 51, N_("HDR")                            },
+        { 55, N_("Handheld Night Shot")            },
+        { 57, N_("3D")                             },
+        { 59, N_("Creative Control")               },
+        { 62, N_("Panorama")                       },
+        { 63, N_("Glass Through")                  },
+        { 64, N_("HDR")                            },
+        { 66, N_("Digital Filter")                 },
+        { 67, N_("Clear Portrait")                 },
+        { 68, N_("Silky Skin")                     },
+        { 69, N_("Backlit Softness")               },
+        { 70, N_("Clear in Backlight")             },
+        { 71, N_("Relaxing Tone")                  },
+        { 72, N_("Sweet Child's Face")             },
+        { 73, N_("Distinct Scenery")               },
+        { 74, N_("Bright Blue Sky")                },
+        { 75, N_("Romantic Sunset Glow")           },
+        { 76, N_("Vivid Sunset Glow")              },
+        { 77, N_("Glistening Water")               },
+        { 78, N_("Clear Nightscape")               },
+        { 79, N_("Cool Night Sky")                 },
+        { 80, N_("Warm Glowing Nightscape")        },
+        { 81, N_("Artistic Nightscape")            },
+        { 82, N_("Glittering Illuminations")       },
+        { 83, N_("Clear Night Portrait")           },
+        { 84, N_("Soft Image of a Flower")         },
+        { 85, N_("Appetizing Food")                },
+        { 86, N_("Cute Dessert")                   },
+        { 87, N_("Freeze Animal Motion")           },
+        { 88, N_("Clear Sports Shot")              },
+        { 89, N_("Monochrome")                     },
+        { 90, N_("Creative Control")               }
     };
 
     //! Audio, tag 0x0020
     extern const TagDetails panasonicAudio[] = {
         { 1, N_("Yes") },
-        { 2, N_("No")  }
+        { 2, N_("No")  },
+        { 3, N_("Stereo")  }
     };
 
     //! ColorEffect, tag 0x0028
@@ -146,27 +208,72 @@ namespace Exiv2 {
         { 2, N_("Warm")            },
         { 3, N_("Cool")            },
         { 4, N_("Black and white") },
-        { 5, N_("Sepia")           }
+        { 5, N_("Sepia")           },
+        { 6, N_("Happy")           },
+        { 8, N_("Vivid")           }
     };
 
     //! BustMode, tag 0x002a
     extern const TagDetails panasonicBurstMode[] = {
-        { 0, N_("Off")              },
-        { 1, N_("Low/High quality") },
-        { 2, N_("Infinite")         }
+        {  0, N_("Off")                          },
+        {  1, N_("On")                           },
+        {  2, N_("Auto Exposure Bracketing")     },
+        {  4, N_("Unlimited")                    },
+        {  8, N_("White Balance Bracketing")     },
+        { 17, N_("On (with flash)")              }
     };
 
     //! Contrast, tag 0x002c
-    extern const TagDetails panasonicContrast[] = {
-        {   0, N_("Normal")      },
-        {   1, N_("Low")         },
-        {   2, N_("High")        },
-        {   6, N_("Medium low")  },
-        {   7, N_("Medium high") },
-        { 256, N_("Low")         },
-        { 272, N_("Standard")    },
-        { 288, N_("High")        },
-        { 288, N_("High")        }              // To silence compiler warning
+    //! Reporting values for current models, older models may differ
+    //! Values 0..10 and 100..138 are based on images from the G6
+    extern const TagDetails panasonicContrastMode[] = {
+        {   0, N_("-5")                                             },
+        {   1, N_("-4")                                             },
+        {   2, N_("-3")                                             },
+        {   3, N_("-2")                                             },
+        {   4, N_("-1")                                             },
+        {   5, N_("Normal")                                         },
+        {   6, N_("+1")                                             },
+        {   7, N_("+2")                                             },
+        {   8, N_("+3")                                             },
+        {   9, N_("+4")                                             },
+        {  10, N_("+5")                                             },
+        {  12, N_("Smooth (Color Film) or Pure (My Color)")         },
+        {  17, N_("Dynamic (B&W Film)")                             },
+        {  22, N_("Smooth (B&W Film)")                              },
+        {  25, N_("High Dynamic")                                   },
+        {  26, N_("Retro")                                          },
+        {  27, N_("Dynamic (Color Film)")                           },
+        {  28, N_("Low Key")                                        },
+        {  29, N_("Toy Effect")                                     },
+        {  32, N_("Vibrant (Color Film) or Expressive (My Color)")  },
+        {  33, N_("Elegant (My Color)")                             },
+        {  37, N_("Nostalgic (Color Film)")                         },
+        {  41, N_("Dynamic Art (My Color)")                         },
+        {  42, N_("Retro (My Color)")                               },
+        {  45, N_("Cinema")                                         },
+        {  47, N_("Dynamic Mono")                                   },
+        {  50, N_("Impressive Art")                                 },
+        {  51, N_("Cross Process")                                  },
+        { 100, N_("High Dynamic")                                   },
+        { 101, N_("Retro")                                          },
+        { 102, N_("High Key")                                       },
+        { 103, N_("Low Key")                                        },
+        { 104, N_("Toy Effect")                                     },
+        { 107, N_("Expressive")                                     },
+        { 112, N_("Sepia")                                          },
+        { 117, N_("Miniature Effect")                               },
+        { 122, N_("Dynamic Monochrome")                             },
+        { 127, N_("Old Days")                                       },
+        { 132, N_("Bleach Pass")                                    },
+        { 135, N_("Impressive Art")                                 },
+        { 136, N_("Cross Process")                                  },
+        { 137, N_("Toy Pop")                                        },
+        { 138, N_("Fantasy")                                        },
+        { 256, N_("Normal 2")                                       },
+        { 272, N_("Standard")                                       },
+        { 288, N_("High")                                           },
+        { 288, N_("High")                                           }              // To silence compiler warning
     };
 
     //! NoiseReduction, tag 0x002d
@@ -180,18 +287,28 @@ namespace Exiv2 {
 
     //! SelfTimer, tag 0x002e
     extern const TagDetails panasonicSelfTimer[] = {
-        { 1, N_("Off") },
-        { 2, N_("10s") },
-        { 3, N_("2s")  }
+        { 1, N_("Off")                  },
+        { 2, N_("10 s")                 },
+        { 3, N_("2 s")                  },
+        { 4, N_("10 s / 3 pictures")    }
     };
 
     //! Rotation, tag 0x0030
     extern const TagDetails panasonicRotation[] = {
         { 1, N_("Horizontal (normal)") },
+        { 3, N_("Rotate 180")          },
         { 6, N_("Rotate 90 CW")        },
         { 8, N_("Rotate 270 CW")       }
     };
 
+    //! AFAssistLamp, tag 0x0031
+    extern const TagDetails panasonicAFAssistLamp[] = {
+        { 1, N_("Fired")                        },
+        { 2, N_("Enabled but Not Used")         },
+        { 3, N_("Disabled but Required")        },
+        { 4, N_("Disabled and Not Required")    }
+    };
+
     //! ColorMode, tag 0x0032
     extern const TagDetails panasonicColorMode[] = {
         { 0, N_("Normal")  },
@@ -202,7 +319,7 @@ namespace Exiv2 {
     //! OpticalZoomMode, tag 0x0034
     extern const TagDetails panasonicOpticalZoomMode[] = {
         { 1, N_("Standard")  },
-        { 2, N_("EX optics") }
+        { 2, N_("Extended") }
     };
 
     //! ConversionLens, tag 0x0035
@@ -222,13 +339,149 @@ namespace Exiv2 {
 
     //! FilmMode, tag 0x0042
     extern const TagDetails panasonicFilmMode[] = {
-        { 1, N_("Standard (color)") },
-        { 2, N_("Dynamic (color)")  },
-        { 3, N_("Nature (color)")   },
-        { 4, N_("Smooth (color)")   },
-        { 5, N_("Standard (B&W)")   },
-        { 6, N_("Dynamic (B&W)")    },
-        { 7, N_("Smooth (B&W)")     }
+        {  1, N_("Standard (color)") },
+        {  2, N_("Dynamic (color)")  },
+        {  3, N_("Nature (color)")   },
+        {  4, N_("Smooth (color)")   },
+        {  5, N_("Standard (B&W)")   },
+        {  6, N_("Dynamic (B&W)")    },
+        {  7, N_("Smooth (B&W)")     },
+        { 10, N_("Nostalgic")        },
+        { 11, N_("Vibrant")          }
+    };
+
+    //! BracketSettings, tag 0x0045
+    extern const TagDetails panasonicBracketSettings[] = {
+        { 0, N_("No Bracket")               },
+        { 1, N_("3 Images, Sequence 0/-/+") },
+        { 2, N_("3 Images, Sequence -/0/+") },
+        { 3, N_("5 Images, Sequence 0/-/+") },
+        { 4, N_("5 Images, Sequence -/0/+") },
+        { 5, N_("7 Images, Sequence 0/-/+") },
+        { 6, N_("7 Images, Sequence -/0/+") }
+    };
+
+    //! FlashCurtain, tag 0x0048
+    extern const TagDetails panasonicFlashCurtain[] = {
+        { 0, N_("n/a")  },
+        { 1, N_("1st")  },
+        { 2, N_("2nd")  }
+    };
+
+    //! LongExposureNoiseReduction, tag 0x0049
+    extern const TagDetails panasonicLongExposureNoiseReduction[] = {
+        { 1, N_("Off")  },
+        { 2, N_("On")  }
+    };
+
+    //! IntelligentExposure, tag 0x005d
+    extern const TagDetails panasonicIntelligentExposure[] = {
+        { 0, N_("Off")      },
+        { 1, N_("Low")      },
+        { 2, N_("Standard") },
+        { 3, N_("High")     }
+    };
+
+    //! FlashWarning, tag 0x0062
+    extern const TagDetails panasonicFlashWarning[] = {
+        { 0, N_("No")  },
+        { 1, N_("Yes (Flash required but disabled)")  }
+    };
+
+    //! IntelligentResolution, tag 0x0070
+    extern const TagDetails panasonicIntelligentResolution[] = {
+        { 0, N_("Off")      },
+        { 1, N_("Low")      },
+        { 2, N_("Standard") },
+        { 3, N_("High")     },
+        { 4, N_("Extended") }
+    };
+
+    //! IntelligentDynamicRange, tag 0x0079
+    extern const TagDetails panasonicIntelligentDynamicRange[] = {
+        { 0, N_("Off")      },
+        { 1, N_("Low")      },
+        { 2, N_("Standard") },
+        { 3, N_("High")     }
+    };
+
+    //! ClearRetouch, tag 0x007c
+    extern const TagDetails panasonicClearRetouch[] = {
+        { 0, N_("Off")  },
+        { 1, N_("On")  }
+    };
+
+    //! PhotoStyle, tag 0x0089
+    extern const TagDetails panasonicPhotoStyle[] = {
+        { 0, N_("Auto")                 },
+        { 1, N_("Standard or Custom")   },
+        { 2, N_("Vivid")                },
+        { 3, N_("Natural")              },
+        { 4, N_("Monochrome")           },
+        { 5, N_("Scenery")              },
+        { 6, N_("Portrait")             }
+    };
+
+    //! ShadingCompensation, tag 0x008a
+    extern const TagDetails panasonicShadingCompensation[] = {
+        { 0, N_("Off")  },
+        { 1, N_("On")  }
+    };
+
+    //! CameraOrientation, tag 0x008f
+    extern const TagDetails panasonicCameraOrientation[] = {
+        { 0, N_("Normal")           },
+        { 1, N_("Rotate CW")        },
+        { 2, N_("Rotate 180")       },
+        { 3, N_("Rotate CCW")       },
+        { 4, N_("Tilt Upwards")     },
+        { 5, N_("Tilt Downwards")   }
+    };
+
+    //! SweepPanoramaDirection, tag 0x0093
+    extern const TagDetails panasonicSweepPanoramaDirection[] = {
+        { 0, N_("Off")              },
+        { 1, N_("Left to Right")    },
+        { 2, N_("Right to Left")    },
+        { 3, N_("Top to Bottom")    },
+        { 4, N_("Bottom to Top")    }
+    };
+
+    //! TimerRecording, tag 0x0096
+    extern const TagDetails panasonicTimerRecording[] = {
+        { 0, N_("Off")                      },
+        { 1, N_("Time Lapse")               },
+        { 2, N_("Stop-motion Animation")    }
+    };
+
+    //! HDR, tag 0x009e
+    extern const TagDetails panasonicHDR[] = {
+        {     0, N_("Off")          },
+        {   100, N_("1 EV")         },
+        {   200, N_("2 EV")         },
+        {   300, N_("3 EV")         },
+        { 32868, N_("1 EV (Auto)")  },
+        { 32968, N_("2 EV (Auto)")  },
+        { 32068, N_("2 EV (Auto)")  }
+    };
+
+    //! ShutterType, tag 0x009f
+    extern const TagDetails panasonicShutterType[] = {
+        { 0, N_("Mechanical")   },
+        { 1, N_("Electronical") },
+        { 2, N_("Hybrid")       }
+    };
+
+    //! TouchAE, tag 0x00ab
+    extern const TagDetails panasonicTouchAE[] = {
+        { 0, N_("Off")  },
+        { 1, N_("On")  }
+    };
+
+    //! FlashFired, tag 0x8007
+    extern const TagDetails panasonicFlashFired[] = {
+        { 1, N_("No")  },
+        { 2, N_("Yes")  }
     };
 
     // Panasonic MakerNote Tag Info
@@ -254,12 +507,12 @@ namespace Exiv2 {
         TagInfo(0x0029, "TimeSincePowerOn", "Time since Power On", N_("Time in 1/100 s from when the camera was powered on to when the image is written to memory card"), panasonicId, makerTags, unsignedLong, -1, printValue),
         TagInfo(0x002a, "BurstMode", N_("Burst Mode"), N_("Burst mode"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicBurstMode)),
         TagInfo(0x002b, "SequenceNumber", N_("Sequence Number"), N_("Sequence number"), panasonicId, makerTags, unsignedLong, -1, printValue),
-        TagInfo(0x002c, "Contrast", N_("Contrast"), N_("Contrast setting"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicContrast)),
+        TagInfo(0x002c, "ContrastMode", N_("Contrast Mode"), N_("Contrast Mode"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicContrastMode)),
         TagInfo(0x002d, "NoiseReduction", N_("NoiseReduction"), N_("Noise reduction"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicNoiseReduction)),
         TagInfo(0x002e, "SelfTimer", N_("Self Timer"), N_("Self timer"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicSelfTimer)),
         TagInfo(0x002f, "0x002f", "0x002f", N_("Unknown"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x0030, "Rotation", N_("Rotation"), N_("Rotation"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicRotation)),
-        TagInfo(0x0031, "0x0031", "0x0031", N_("Unknown"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0031, "AFAssistLamp", "AF Assist Lamp", N_("Auto Focus Assist Lamp"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicAFAssistLamp)),
         TagInfo(0x0032, "ColorMode", N_("Color Mode"), N_("Color mode"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicColorMode)),
         TagInfo(0x0033, "BabyAge", N_("Baby Age"), N_("Baby (or pet) age"), panasonicId, makerTags, asciiString, -1, printValue),
         TagInfo(0x0034, "OpticalZoomMode", N_("Optical Zoom Mode"), N_("Optical zoom mode"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicOpticalZoomMode)),
@@ -268,14 +521,52 @@ namespace Exiv2 {
         TagInfo(0x0039, "Contrast", N_("Contrast"), N_("Contrast"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x003a, "WorldTimeLocation", N_("World Time Location"), N_("World time location"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicWorldTimeLocation)),
         TagInfo(0x003c, "ProgramISO", N_("Program ISO"), N_("Program ISO"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x003d, "AdvancedSceneType", N_("Advanced Scene Type"), N_("Advanced Scene Type"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x003f, "FacesDetected", N_("Faces Detected"), N_("Faces Detected"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x0040, "Saturation", N_("Saturation"), N_("Saturation"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x0041, "Sharpness", N_("Sharpness"), N_("Sharpness"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x0042, "FilmMode", N_("Film Mode"), N_("Film mode"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicFilmMode)),
+        TagInfo(0x0044, "ColorTempKelvin", N_("Color Temperature (Kelvin)"), N_("Color Temperature (Kelvin)"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0045, "BracketSettings", N_("Auto Bracketing Settings"), N_("Auto Bracketing Settings"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicBracketSettings)),
         TagInfo(0x0046, "WBAdjustAB", N_("WB Adjust AB"), N_("WB adjust AB. Positive is a shift toward blue."), panasonicId, makerTags, unsignedShort, -1, printValue),
-        TagInfo(0x0047, "WBAdjustGM", N_("WB Adjust GM"), N_("WBAdjustGM. Positive is a shift toward green."), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0047, "WBAdjustGM", N_("WB Adjust GM"), N_("WB adjust GM. Positive is a shift toward green."), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0048, "FlashCurtain", N_("Flash Curtain"), N_("Flash Curtain"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicFlashCurtain)),
+        TagInfo(0x0049, "LongExposureNoiseReduction", N_("Long Exposure Noise Reduction"), N_("Long Exposure Noise Reduction"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicLongExposureNoiseReduction)),
         TagInfo(0x0051, "LensType", N_("Lens Type"), N_("Lens type"), panasonicId, makerTags, asciiString, -1, printValue),
         TagInfo(0x0052, "LensSerialNumber", N_("Lens Serial Number"), N_("Lens serial number"), panasonicId, makerTags, asciiString, -1, printValue),
         TagInfo(0x0053, "AccessoryType", N_("Accessory Type"), N_("Accessory type"), panasonicId, makerTags, asciiString, -1, printValue),
+        TagInfo(0x0054, "AccessorySerialNumber", N_("Accessory Serial Number"), N_("Accessory Serial Number"), panasonicId, makerTags, asciiString, -1, printValue),
+        TagInfo(0x005d, "IntelligentExposure", N_("Intelligent Exposure"), N_("Intelligent Exposure"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicIntelligentExposure)),
+        TagInfo(0x0060, "LensFirmwareVersion", N_("Lens Firmware Version"), N_("Lens Firmware Version"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x0062, "FlashWarning", N_("Flash Warning"), N_("Flash Warning"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicFlashWarning)),
+        TagInfo(0x0065, "Title", N_("Title"), N_("Title"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x0066, "BabyName", N_("Baby Name"), N_("Baby Name"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x0067, "Location", N_("Location"), N_("Location"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x0069, "Country", N_("Country"), N_("Country"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x006b, "State", N_("State"), N_("State"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x006d, "City", N_("City"), N_("City"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x006f, "Landmark", N_("Landmark"), N_("Landmark"), panasonicId, makerTags, undefined, -1, printValue),
+        TagInfo(0x0070, "IntelligentResolution", N_("Intelligent Resolution"), N_("Intelligent Resolution"), panasonicId, makerTags, unsignedByte, -1, EXV_PRINT_TAG(panasonicIntelligentResolution)),
+        TagInfo(0x0077, "BurstSpeed", N_("Burst Speed"), N_("Burst Speed (Images per Second"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0079, "IntelligentDynamicRange", N_("Intelligent Dynamic Range"), N_("Intelligent Dynamic Range"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicIntelligentDynamicRange)),
+        TagInfo(0x007c, "ClearRetouch", N_("Clear Retouch"), N_("Clear Retouch"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicClearRetouch)),
+        TagInfo(0x0086, "ManometerPressure", N_("Manometer Pressure"), N_("Manometer Pressure"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0089, "PhotoStyle", N_("Photo Style"), N_("Photo Style"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicPhotoStyle)),
+        TagInfo(0x008a, "ShadingCompensation", N_("Shading Compensation"), N_("Shading Compensation"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicShadingCompensation)),
+        TagInfo(0x008c, "AccelerometerZ", N_("Accelerometer Z"), N_("Accelerometer Z (Positive is acceleration upwards"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x008d, "AccelerometerX", N_("Accelerometer X"), N_("Accelerometer X (Positive is acceleration to the left"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x008e, "AccelerometerY", N_("Accelerometer Y"), N_("Accelerometer Y (Positive is acceleration backwards)"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x008f, "Camera Orientation", N_("Camera Orientation"), N_("Camera Orientation"), panasonicId, makerTags, unsignedByte, -1, EXV_PRINT_TAG(panasonicCameraOrientation)),
+        TagInfo(0x0090, "RollAngle", N_("Roll Angle"), N_("Roll Angle (Clockwise camera rotation in degrees)"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0091, "PitchAngle", N_("Pitch Angle"), N_("Pitch Angle (Upward camera tilt in degrees)"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0093, "SweepPanoramaDirection", N_("Sweep Panorama Direction"), N_("Sweep Panorama Direction"), panasonicId, makerTags, unsignedByte, -1, EXV_PRINT_TAG(panasonicSweepPanoramaDirection)),
+        TagInfo(0x0094, "SweepPanoramaFieldOfView", N_("Sweep Panorama Field of View"), N_("Sweep Panorama Field of View"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x0096, "TimerRecording", N_("Timer Recording"), N_("Timer Recording"), panasonicId, makerTags, unsignedByte, -1, EXV_PRINT_TAG(panasonicTimerRecording)),
+        TagInfo(0x009d, "InternalNDFilter", N_("Internal ND Filter"), N_("Internal HD Filter"), panasonicId, makerTags, unsignedRational, -1, printValue),
+        TagInfo(0x009e, "HDR", N_("High Dynamic Range"), N_("High Dynamic Range"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicHDR)),
+        TagInfo(0x009f, "ShutterType", N_("Shutter Type"), N_("Shutter Type"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicShutterType)),
+        TagInfo(0x00a3, "ClearRetouchValue", N_("Clear Retouch Value"), N_("Clear Retouch Value"), panasonicId, makerTags, unsignedRational, -1, printValue),
+        TagInfo(0x00ab, "TouchAE", N_("Touch AE"), N_("Touch Auto Focus and Exposure"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicTouchAE)),
         TagInfo(0x0e00, "PrintIM", N_("Print IM"), N_("PrintIM information"), panasonicId, makerTags, undefined, -1, printValue),
         TagInfo(0x4449, "0x4449", "0x4449", N_("Unknown"), panasonicId, makerTags, undefined, -1, printValue),
         TagInfo(0x8000, "MakerNoteVersion", N_("MakerNote Version"), N_("MakerNote version"), panasonicId, makerTags, undefined, -1, printExifVersion),
@@ -283,6 +574,7 @@ namespace Exiv2 {
         TagInfo(0x8004, "WBRedLevel", N_("WB Red Level"), N_("WB red level"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x8005, "WBGreenLevel", N_("WB Green Level"), N_("WB green level"), panasonicId, makerTags, unsignedShort, -1, printValue),
         TagInfo(0x8006, "WBBlueLevel", N_("WB Blue Level"), N_("WB blue level"), panasonicId, makerTags, unsignedShort, -1, printValue),
+        TagInfo(0x8007, "FlashFired", N_("Flash Fired"), N_("Flash Fired"), panasonicId, makerTags, unsignedShort, -1, EXV_PRINT_TAG(panasonicFlashFired)),
         TagInfo(0x8010, "BabyAge", N_("Baby Age"), N_("Baby (or pet) age"), panasonicId, makerTags, asciiString, -1, printValue),
         // End of list marker
         TagInfo(0xffff, "(UnknownPanasonicMakerNoteTag)", "(UnknownPanasonicMakerNoteTag)", N_("Unknown PanasonicMakerNote tag"), panasonicId, makerTags, asciiString, -1, printValue)
@@ -302,16 +594,18 @@ namespace Exiv2 {
         }
         long l0 = value.toLong(0);
         long l1 = value.toLong(1);
-        if      (l0 ==  0 && l1 ==  1) os << _("Spot mode on");
-        else if (l0 ==  0 && l1 == 16) os << _("Spot mode off or 3-area (high speed)");
-        else if (l0 ==  1 && l1 ==  0) os << _("Spot focussing");
-        else if (l0 ==  1 && l1 ==  1) os << _("5-area");
-        else if (l0 == 16 && l1 ==  0) os << _("1-area");
-        else if (l0 == 16 && l1 == 16) os << _("1-area (high speed)");
-        else if (l0 == 32 && l1 ==  0) os << _("3-area (auto)");
-        else if (l0 == 32 && l1 ==  1) os << _("3-area (left)");
-        else if (l0 == 32 && l1 ==  2) os << _("3-area (center)");
-        else if (l0 == 32 && l1 ==  3) os << _("3-area (right)");
+        if      (l0 ==   0 && l1 ==  1) os << _("Spot mode on");
+        else if (l0 ==   0 && l1 == 16) os << _("Spot mode off or 3-area (high speed)");
+        else if (l0 ==   1 && l1 ==  0) os << _("Spot Focussing");
+        else if (l0 ==   1 && l1 ==  1) os << _("5-area");
+        else if (l0 ==  16 && l1 ==  0) os << _("1-area");
+        else if (l0 ==  16 && l1 == 16) os << _("1-area (high speed)");
+        else if (l0 ==  32 && l1 ==  0) os << _("3-area (auto)");
+        else if (l0 ==  32 && l1 ==  1) os << _("3-area (left)");
+        else if (l0 ==  32 && l1 ==  2) os << _("3-area (center)");
+        else if (l0 ==  32 && l1 ==  3) os << _("3-area (right)");
+        else if (l0 ==  64 && l1 ==  0) os << _("Face Detect");
+        else if (l0 == 128 && l1 ==  0) os << _("Spot Focussing 2");
         else os << value;
         return os;
     } // PanasonicMakerNote::print0x000f
@@ -346,6 +640,8 @@ namespace Exiv2 {
         TagInfo(0x0025, "WBGreenLevel", N_("WB Green Level"), N_("WB green level"), panaRawId, panaRaw, unsignedShort, -1, printValue),
         TagInfo(0x0026, "WBBlueLevel", N_("WB Blue Level"), N_("WB blue level"), panaRawId, panaRaw, unsignedShort, -1, printValue),
         TagInfo(0x002e, "PreviewImage", N_("Preview Image"), N_("Preview image"), panaRawId, panaRaw, undefined, -1, printValue),
+        TagInfo(0x004b, "PanasonicImageWidth", N_("Panasonic Image Width"), N_("Panasonic Image Width"), panasonicId, makerTags, unsignedLong, -1, printValue),
+        TagInfo(0x004c, "PanasonicImageHeight", N_("Panasonic Image Height"), N_("Panasonic Image Height"), panasonicId, makerTags, unsignedLong, -1, printValue),
         TagInfo(0x010f, "Make", N_("Manufacturer"), N_("The manufacturer of the recording equipment"), panaRawId, panaRaw, asciiString, -1, printValue),
         TagInfo(0x0110, "Model", N_("Model"), N_("The model name or model number of the equipment"), panaRawId, panaRaw, asciiString, -1, printValue),
         TagInfo(0x0111, "StripOffsets", N_("Strip Offsets"), N_("Strip offsets"), panaRawId, panaRaw, unsignedLong, -1, printValue),
-- 
2.0.1

